File: src/models/lore/Organization.ts
import Lore, {LoreType} from "./Lore";

export enum OrgType {
    Religious = 'Religious',
    Social = 'Social',
    Political = 'Political',
    Military = 'Military',
    Academic = 'Academic'
}

export interface Organization extends Lore {
    orgType: OrgType
    founded: number
    disbanded: number
    nickname: string
    membersName: string
}

export enum OrgKey {
    name = 'name',
    type = 'type',
    orgType = 'orgType',
    founded = 'founded',
    disbanded = 'disbanded',
    nickname = 'nickname',
    membersName = 'membersName'
}

export class DefaultOrganization {
    static create(): Organization {
        return {
            name: '',
            type: LoreType.ORGANIZATION,
            orgType: OrgType.Religious,
            founded: 0,
            disbanded: 0,
            nickname: '',
            membersName: ''
        }
    }
}

File: server/src/main/java/com/github/genraven/genesys/domain/lore/Organization.java
package com.github.genraven.genesys.domain.lore;

import com.fasterxml.jackson.annotation.JsonValue;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import org.springframework.data.mongodb.core.mapping.Document;

@EqualsAndHashCode(callSuper = true)
@Data
@Document(collection = "organizations")
public class Organization extends Lore {

    protected Organization() {}

    public Organization(final Lore lore) {
        this.setId(lore.getId());
        this.setName(lore.getName());
        this.setType(Type.ORGANIZATION);
    }

    private OrgType orgType;
    private int founded;
    private int disbanded;
    private String nickname;
    private String membersName;

    @Getter
    @AllArgsConstructor
    public enum OrgType {
        POLITICAL("Political"),
        SOCIAL("Social"),
        RELIGIOUS("Religious"),
        MILITARY("Military"),
        ACADEMIC("Academic");

        @JsonValue
        private final String label;
    }
}
